OpenWeather/Networking/Implementation/LocationProvider.swift

```diff
@@ -55,10 +54,76 @@ struct LocationProvider: LocationProviding {
             }
         }
     }
+
+    func locations(for address: String) async throws -> [DeviceLocation] {
+        let response: LocationSearchResponse = try await locations(for: address)
+
+        return response.result
+            .map {
+                DeviceLocation(
+                    name: $0.name,
+                    location: $0.location
+                )
+            }
+    }
+
+    private func locations(for address: String) async throws -> LocationSearchResponse {
+        return try await withCheckedThrowingContinuation { continuation in
+            locations(
+                for: address,
+                completion: { placemarks, error in
+                    if let error = error {
+                        continuation.resume(
+                            throwing: LocationProvidingError.failure(reason: error.localizedDescription)
+                        )
+                    } else {
+                        let result: [NetworkLocation] = placemarks
+                            .compactMap { placemark -> NetworkLocation? in
+                                guard
+                                    let name = placemark.name,
+                                    let location = placemark.location
+                                else { return nil }
+
+                                return NetworkLocation(
+                                    name: name,
+                                    location: location
+                                )
+                            }
+
+                        continuation.resume(
+                            returning: LocationSearchResponse(result: result)
+                        )
+                    }
+                }
+            )
+        }
+    }
+
+    private func locations(
+        for address: String,
+        completion: @escaping ([CLPlacemark], Error?) -> Void
+    ) {
+        geocoder.geocodeAddressString(address) { placemarks, error in
+            if let error = error as? NSError {
+                completion([], error)
+            } else {
+                completion(placemarks ?? [], nil)
+            }
+        }
+    }
 }
 
 struct MockLocationProvider: LocationProviding {
     func locationName(for location: CLLocation) async throws -> String {
         return Mock.locationNameMap[location] ?? "Unknown"
     }
+
+    func locations(for address: String) async throws -> [DeviceLocation] {
+        guard address.isEmpty == false else { return [] }
+
+        return Mock.locationsMap
+            .compactMap { (key, value) -> DeviceLocation? in
+                return key.lowercased().starts(with: address.lowercased()) ? value : nil
```

Comment:
`starts` is fine, but maybe we use `contains`? Doesn't really matter as it is a mock service, but wanted to suggest it as I normally use `contains`. Up to you!